psi4/src/psi4/libmints/thc_eri.cc

```diff
@@ -259,6 +259,51 @@ SharedMatrix LS_THC_Computer::build_E_df() {
     return linalg::triplet(E_temp, Jinv, E_temp, false, false, true);
 }
 
+/// @brief Drastically reduces the size of the grid using pivoted Cholesky decomposition (Parrish et al. 2012 procedure 2)
+SharedMatrix LS_THC_Computer::prune_grid() {
+
+    outfile->Printf("    Pruning grid using pivoted Cholesky (Matthews 2020)...\n");
+
+    size_t npoints = x1_->nrow();
+    size_t nbf = x1_->ncol();
+
+    SharedMatrix S_Qq = std::make_shared<Matrix>(npoints, npoints);
+    SharedMatrix S_temp = linalg::doublet(x1_, x1_, false, true);
+
+    auto S_Qq_p = S_Qq->pointer();
+    auto Stp = S_temp->pointer();
+
+    // Parrish 2012, Equation 28
+#pragma omp parallel for
+    for (size_t p = 0; p < npoints; ++p) {
+        for (size_t q = 0; q < npoints; ++q) {
+            S_Qq_p[p][q] = Stp[p][q] * Stp[p][q];
+        }
+    }
+
+    std::vector<std::vector<int>> pivot;
+    S_Qq->pivoted_cholesky(PSI_ZERO, pivot);
+
+    // Update S_Qq (returned as upper Cholesky factor of dimension (npoints, rank))
+    S_Qq = linalg::doublet(S_Qq, S_Qq, true, false);
+    size_t rank = S_Qq->nrow();
+
+    // Update factor matrix
+    auto x1_new = std::make_shared<Matrix>(rank, nbf);
+#pragma omp parallel for
```

Comment:
You could probably use the above-suggested `collapse(2) schedule(guided)` strategy on this loop, as well.